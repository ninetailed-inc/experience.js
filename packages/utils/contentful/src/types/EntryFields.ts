import { z } from 'zod';

/**
 * This does not work anymore from zod > 3.21.0:
 *
 * We have to cast the result of passthrough() to z.ZodObject<{}> to make it work,
 * as the inferred type changed to {} & { [k: string]: unknown; }
 * It was {} before, so we do the type cast to get it back to {}.
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export const EntryFields = z.object({}).passthrough() as z.ZodObject<{}>;
export type EntryFields = z.infer<typeof EntryFields>;
