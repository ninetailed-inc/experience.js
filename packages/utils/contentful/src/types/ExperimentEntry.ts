import { SafeParseError, SafeParseSuccess, z } from 'zod';
import {
  ExperienceEntrySchema,
  ExperienceEntry,
  ExperienceEntryFields,
  ExperienceEntryLike,
} from './ExperienceEntry';
import { EntryFields } from './EntryFields';

export const ExperimentEntrySchema = ExperienceEntrySchema.extend({
  fields: ExperienceEntryFields.extend({
    nt_type: z.string().regex(/^nt_experiment$/g),
  }),
});

const parse = <T extends EntryFields>(
  input: ExperienceEntryLike<T>
): ExperienceEntry<T> => {
  const output = ExperimentEntrySchema.parse(input);

  return {
    ...output,
    fields: { ...output.fields, nt_variants: input.fields.nt_variants || [] },
  };
};

export const safeParse = <T extends EntryFields>(
  input: ExperienceEntryLike<T>
):
  | SafeParseError<ExperienceEntryLike<T>>
  | SafeParseSuccess<ExperienceEntry<T>> => {
  const output = ExperimentEntrySchema.safeParse(input);

  if (!output.success) {
    return output;
  }

  return {
    ...output,
    data: {
      ...output.data,
      fields: {
        ...output.data.fields,
        nt_variants: input.fields.nt_variants || [],
      },
    },
  };
};

export const ExperimentEntry = {
  ...ExperimentEntrySchema,
  parse,
  safeParse,
};

export type ExperimentEntryLike = z.input<typeof ExperimentEntrySchema>;
export type ExperimentEntry = z.infer<typeof ExperimentEntrySchema>;
