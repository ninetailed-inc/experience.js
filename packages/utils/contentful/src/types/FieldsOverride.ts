import { EntryFields } from './EntryFields';

export type FieldsOverride<Entry, Fields extends EntryFields> = Omit<
  Entry,
  'fields'
> & {
  fields: Fields;
};
