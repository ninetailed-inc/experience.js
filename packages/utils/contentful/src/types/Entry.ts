import { z } from 'zod';

import { EntryFields } from './EntryFields';
import { FieldsOverride } from './FieldsOverride';

const EntryLink = z.object({
  type: z.string().optional(),
  linkType: z.string().optional(),
  id: z.string(),
});

const LinkedEntity = z.object({
  sys: EntryLink,
});

export const EntrySchema = z.object({
  sys: z.object({
    type: z.string().optional(),
    id: z.string(),
    createdAt: z.string().optional(),
    updatedAt: z.string().optional(),
    locale: z.string().optional(),
    revision: z.number().optional(),
    space: LinkedEntity.optional(),
    environment: LinkedEntity.optional(),
    contentType: LinkedEntity.optional(),
  }),
  fields: EntryFields,
  metadata: z
    .object({
      tags: z.array(
        z.object({
          sys: EntryLink.extend({ linkType: z.string() }),
        })
      ),
    })
    .optional(),
});

export type EntryLike<T extends EntryFields = EntryFields> = FieldsOverride<
  z.input<typeof EntrySchema>,
  T
>;

export type Entry<T extends EntryFields = EntryFields> = FieldsOverride<
  z.infer<typeof EntrySchema>,
  T
>;

const parse = <T extends EntryFields>(input: EntryLike<T>): Entry<T> => {
  const output = EntrySchema.parse(input);

  return {
    ...output,
    fields: input.fields,
  };
};

const safeParse = <T extends EntryFields>(input: EntryLike<T>) => {
  const output = EntrySchema.safeParse(input);

  if (!output.success) {
    return output;
  }

  return {
    ...output,
    data: {
      ...output.data,
      fields: input.fields,
    },
  };
};

export const Entry = {
  ...EntrySchema,
  parse,
  safeParse,
};
