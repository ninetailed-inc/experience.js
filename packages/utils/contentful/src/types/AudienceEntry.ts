import { z } from 'zod';
import { EntrySchema } from './Entry';
import { EntryFields } from './EntryFields';

export const AudienceEntryFields = EntryFields.extend({
  /**
   * The name of the audience (Short Text)
   */
  nt_name: z.string(),

  /**
   * The description of the audience (Short Text)
   */
  nt_description: z.string().optional(),

  /**
   * The internal id of the audience (Short Text)
   */
  nt_audience_id: z.string(),
});
export type AudienceEntryFields = z.infer<typeof AudienceEntryFields>;

export const AudienceEntry = EntrySchema.extend({
  fields: AudienceEntryFields,
});

export type AudienceEntryLike = z.input<typeof AudienceEntry>;
export type AudienceEntry = z.infer<typeof AudienceEntry>;
