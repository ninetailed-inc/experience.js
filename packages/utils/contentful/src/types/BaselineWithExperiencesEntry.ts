import { z } from 'zod';

import { EntrySchema } from './Entry';
import { EntryFields } from './EntryFields';
import {
  ExperienceEntry,
  ExperienceEntryLike,
  ExperienceEntrySchema,
} from './ExperienceEntry';
import { FieldsOverride } from './FieldsOverride';

const BaselineWithExperiencesEntryFields = EntryFields.extend({
  nt_experiences: z.array(ExperienceEntrySchema),
});

export type BaselineWithExperiencesEntryFieldsLike<
  VariantFields extends EntryFields = EntryFields
> = Omit<
  z.input<typeof BaselineWithExperiencesEntryFields>,
  'nt_experiences'
> & {
  nt_experiences?: ExperienceEntryLike<VariantFields>[];
};

export type BaselineWithExperiencesEntryFields<
  VariantFields extends EntryFields = EntryFields
> = Omit<
  z.infer<typeof BaselineWithExperiencesEntryFields>,
  'nt_experiences'
> & { nt_experiences: ExperienceEntryLike<VariantFields>[] };

export const BaselineWithExperiencesEntrySchema = EntrySchema.extend({
  fields: BaselineWithExperiencesEntryFields,
});

type BaselineWithExperiencesEntryInput = z.input<
  typeof BaselineWithExperiencesEntrySchema
>;
export type BaselineWithExperiencesEntryLike<
  VariantFields extends EntryFields = EntryFields
> = FieldsOverride<
  BaselineWithExperiencesEntryInput,
  BaselineWithExperiencesEntryFieldsLike<VariantFields>
>;

type BaselineWithExperiencesEntryOutput = z.infer<
  typeof BaselineWithExperiencesEntrySchema
>;
export type BaselineWithExperiencesEntry<
  VariantFields extends EntryFields = EntryFields
> = Omit<BaselineWithExperiencesEntryOutput, 'nt_experiences'> & {
  nt_experiences: ExperienceEntry<VariantFields>[];
};
