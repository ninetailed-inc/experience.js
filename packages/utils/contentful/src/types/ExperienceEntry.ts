import { SafeParseError, SafeParseSuccess, z } from 'zod';
import { Config, ConfigLike } from '@ninetailed/experience.js-utils';

import { EntrySchema, EntryLike, Entry } from './Entry';
import { AudienceEntry } from './AudienceEntry';
import { EntryFields } from './EntryFields';
import { FieldsOverride } from './FieldsOverride';

export const ExperienceEntryFields = z.object({
  /**
   * The name of the experience (Short Text)
   */
  nt_name: z.string(),

  /**
   * The description of the experience (Short Text)
   */
  nt_description: z.string().optional().nullable(),

  /**
   * The type if the experience (nt_experiment | nt_personalization)
   */
  nt_type: z.union([z.string(), z.string()]),

  /**
   * The config of the experience (JSON)
   */
  nt_config: Config.optional()
    .nullable()
    .default(null)
    .transform((val) => {
      return (
        val ?? {
          traffic: 0,
          distribution: [0.5, 0.5],
          components: [],
          sticky: false,
        }
      );
    }),

  /**
   * The audience of the experience (Audience)
   */
  nt_audience: AudienceEntry.optional().nullable(),

  /**
   * All used variants of the experience (Contentful references to other Content Types)
   */
  nt_variants: z.array(EntrySchema).optional(),
});
export type ExperienceFieldsLike<
  VariantFields extends EntryFields = EntryFields
> = Omit<z.infer<typeof ExperienceEntryFields>, 'nt_variants' | 'nt_config'> & {
  nt_variants?: EntryLike<VariantFields>[];
  nt_config?: ConfigLike;
};
export type ExperienceFields<VariantFields extends EntryFields = EntryFields> =
  Omit<z.infer<typeof ExperienceEntryFields>, 'nt_variants'> & {
    nt_variants: Entry<VariantFields>[];
  };

export const ExperienceEntrySchema = EntrySchema.extend({
  fields: ExperienceEntryFields,
});

type ExperienceEntryInput = z.input<typeof ExperienceEntrySchema>;
export type ExperienceEntryLike<
  VariantFields extends EntryFields = EntryFields
> = FieldsOverride<ExperienceEntryInput, ExperienceFieldsLike<VariantFields>>;

type ExperienceEntryOutput = z.infer<typeof ExperienceEntrySchema>;
export type ExperienceEntry<VariantFields extends EntryFields = EntryFields> =
  FieldsOverride<ExperienceEntryOutput, ExperienceFields<VariantFields>>;

const parse = <T extends EntryFields>(
  input: ExperienceEntryLike<T>
): ExperienceEntry<T> => {
  const output = ExperienceEntrySchema.parse(input);

  return {
    ...output,
    fields: { ...output.fields, nt_variants: input.fields.nt_variants || [] },
  };
};

const safeParse = <T extends EntryFields>(
  input: ExperienceEntryLike<T>
):
  | SafeParseError<ExperienceEntryLike<T>>
  | SafeParseSuccess<ExperienceEntry<T>> => {
  const output = ExperienceEntrySchema.safeParse(input);

  if (!output.success) {
    return output;
  }

  return {
    ...output,
    data: {
      ...output.data,
      fields: {
        ...output.data.fields,
        nt_variants: input.fields.nt_variants || [],
      },
    },
  };
};

export const ExperienceEntry = {
  ...ExperienceEntrySchema,
  parse,
  safeParse,
};
