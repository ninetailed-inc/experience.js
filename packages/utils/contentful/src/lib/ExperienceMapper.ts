import { Reference } from '@ninetailed/experience.js';
import { logger } from '@ninetailed/experience.js-shared';
import {
  Audience,
  Experience,
  ExperienceMapper as DefaultExperienceMapper,
} from '@ninetailed/experience.js-utils';

import { AudienceEntry } from '../types/AudienceEntry';
import { BaselineWithExperiencesEntry } from '../types/BaselineWithExperiencesEntry';
import { EntryLike } from '../types/Entry';
import type { EntryFields } from '../types/EntryFields';
import { ExperienceEntry, ExperienceEntryLike } from '../types/ExperienceEntry';
import type { ExperienceFields } from '../types/ExperienceEntry';
import { ExperimentEntry } from '../types/ExperimentEntry';

function mapAudience(ctfAudienceEntry: AudienceEntry): Audience {
  return {
    id: ctfAudienceEntry.fields.nt_audience_id,
    name: ctfAudienceEntry.fields.nt_name,
    description: ctfAudienceEntry.fields.nt_description || '',
  };
}

function createExperience<Variant extends Reference>(
  id: string,
  fields: ExperienceFields,
  variants: Variant[]
): Experience<Variant> {
  const { nt_name, nt_description, nt_type, nt_audience, nt_config } = fields;
  return {
    id,
    name: nt_name,
    description: nt_description || '',
    type: nt_type,
    ...(nt_audience ? { audience: mapAudience(nt_audience) } : {}),
    config: nt_config,
    variants,
  };
}

function validateExperienceEntry<VariantFields extends EntryFields>(
  entry: ExperienceEntryLike<VariantFields>
): ExperienceEntry<VariantFields> {
  const parsedExperience = ExperienceEntry.safeParse(entry);
  if (!parsedExperience.success) {
    logger.warn(
      '[Ninetailed Contentful ExperienceMapper]',
      'Error parsing experience',
      parsedExperience.error.format()
    );
    throw new Error(
      `[Ninetailed Contentful ExperienceMapper] The Experience Input is not valid. Please filter data first with "ExperienceMapper.isExperienceEntry".\n${JSON.stringify(
        parsedExperience.error.format(),
        null,
        2
      )}`
    );
  }
  return parsedExperience.data;
}

export type MapVariantFunction<
  In extends EntryFields,
  Out extends Reference
> = (input: EntryLike<In>) => Out;

export class ExperienceMapper {
  static isExperienceEntry<VariantFields extends EntryFields>(
    entry: ExperienceEntryLike<VariantFields>
  ): entry is ExperienceEntry<VariantFields> {
    return ExperienceEntry.safeParse(entry).success;
  }

  static mapExperience<VariantFields extends EntryFields>(
    ctfEntry: ExperienceEntryLike<VariantFields>
  ) {
    const { sys, fields } = validateExperienceEntry(ctfEntry);
    const variants = fields.nt_variants.map((variant) => ({
      ...variant,
      id: variant.sys.id,
    }));
    const experience = createExperience(sys.id, fields, variants);
    return DefaultExperienceMapper.mapExperience(experience);
  }

  static mapCustomExperience<
    Variant extends Reference,
    VariantFields extends EntryFields
  >(
    ctfEntry: ExperienceEntryLike<VariantFields>,
    mapFn: MapVariantFunction<VariantFields, Variant>
  ) {
    const { sys, fields } = validateExperienceEntry(ctfEntry);
    const variants = fields.nt_variants.map(mapFn);
    const experience = createExperience(sys.id, fields, variants);
    return DefaultExperienceMapper.mapExperience(experience);
  }

  static isExperiment(entry: ExperienceEntryLike): entry is ExperimentEntry {
    return ExperimentEntry.safeParse(entry).success;
  }

  static mapExperiment(entry: ExperienceEntryLike) {
    return ExperienceMapper.mapCustomExperience(entry, () => ({
      id: '',
    }));
  }

  static mapBaselineWithExperiences<Fields extends EntryFields>(
    entry: BaselineWithExperiencesEntry<Fields>
  ) {
    return entry.fields.nt_experiences
      .filter(ExperienceMapper.isExperienceEntry)
      .map((experience) => ExperienceMapper.mapExperience(experience));
  }
}
