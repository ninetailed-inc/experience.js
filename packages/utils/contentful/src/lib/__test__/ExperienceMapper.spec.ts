import { ExperienceEntryLike } from '../../types/ExperienceEntry';
import { ExperienceMapper } from '../ExperienceMapper';
import { EXPERIENCES } from './ntExperiences';
import { ctfHero } from './ctfHero';

interface IHeroFields {
  headline: string;
}

interface IButtonFields {
  label: string;
}

describe('Contentful Experience Mapper', () => {
  it('should map experiences', () => {
    const experienceConfig = EXPERIENCES.map((experience) => {
      return ExperienceMapper.mapCustomExperience(experience, (variant) => ({
        id: variant.sys.id,
        headline: variant.fields.entryTitle,
        x: variant.fields.desktopImage,
      }));
    });

    expect(experienceConfig).toMatchSnapshot();
  });

  it('mapExperience should not have any undefined values', () => {
    const experienceConfig = EXPERIENCES.map((experience) => {
      return ExperienceMapper.mapExperience(experience);
    });

    expect(experienceConfig[0].description).not.toBeUndefined();
    expect(experienceConfig[0].name).not.toBeUndefined();
    expect(experienceConfig[0].audience?.description).not.toBeUndefined();
    expect(experienceConfig[0].audience?.name).not.toBeUndefined();
  });

  it('mapCustomExperience should not have any undefined values', () => {
    const experienceConfig = EXPERIENCES.map((experience) => {
      return ExperienceMapper.mapCustomExperience(experience, (variant) => ({
        id: variant.sys.id,
        headline: variant.fields.entryTitle,
        x: variant.fields.desktopImage,
      }));
    });

    expect(experienceConfig[0].description).not.toBeUndefined();
    expect(experienceConfig[0].name).not.toBeUndefined();
    expect(experienceConfig[0].audience?.description).not.toBeUndefined();
    expect(experienceConfig[0].audience?.name).not.toBeUndefined();
  });

  it('should map experiences with Variant Interfaces', () => {
    const ntExperiences: ExperienceEntryLike<IHeroFields | IButtonFields>[] =
      [];

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const mapped = ntExperiences
      .filter(ExperienceMapper.isExperienceEntry)
      .map((experience) => ExperienceMapper.mapExperience(experience));
  });

  it('should find experiments', () => {
    const ntExperiences: ExperienceEntryLike<IHeroFields | IButtonFields>[] =
      [];

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const mapped = ntExperiences
      .filter(ExperienceMapper.isExperienceEntry)
      .filter(ExperienceMapper.isExperiment)
      .map((experience) => ExperienceMapper.mapExperience(experience));
  });

  it('should work with autogenerated types from contentful', () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const mapped = (ctfHero.fields.nt_experiences || [])
      .filter((entry) => ExperienceMapper.isExperienceEntry(entry))
      .filter(ExperienceMapper.isExperiment)
      .map((experience) => ExperienceMapper.mapExperience(experience));
  });

  it('Should map experiences with custom mapping function for variants including an async step', async () => {
    const mapped = await Promise.all(
      EXPERIENCES.map((experience) =>
        ExperienceMapper.mapCustomExperienceAsync(
          experience,
          async (variant) => {
            await new Promise((resolve) => setTimeout(resolve, 1000));
            return {
              id: variant.sys.id,
              headline: variant.fields.entryTitle,
              image: variant.fields.desktopImage,
            };
          }
        )
      )
    );

    mapped.forEach((mappedExperience) => {
      expect(mappedExperience.description).not.toBeUndefined();
      expect(mappedExperience.name).not.toBeUndefined();
      expect(mappedExperience.audience?.description).not.toBeUndefined();
      expect(mappedExperience.audience?.name).not.toBeUndefined();
      mappedExperience.components.forEach((component) => {
        component.variants.forEach((variant) => {
          expect(variant.id).not.toBeUndefined();

          if ('headline' in variant) {
            expect(variant.headline).not.toBeUndefined();
            expect(variant.image).not.toBeUndefined();
          }
        });
      });
    });
  });
});
