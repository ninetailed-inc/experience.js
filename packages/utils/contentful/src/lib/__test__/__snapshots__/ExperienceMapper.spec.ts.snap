// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contentful Experience Mapper should map experiences 1`] = `
Array [
  Object {
    "audience": Object {
      "description": "",
      "id": "6a26biS02WxlM4WqTiwEZm",
      "name": "Is Cool",
    },
    "components": Array [
      Object {
        "baseline": Object {
          "hidden": false,
          "id": "2t0zGNkAgnSyhUAFigL1Fe",
        },
        "type": "EntryReplacement",
        "variants": Array [
          Object {
            "headline": "[Ninetailed Variant] Home Page Hero Banner",
            "id": "6Vlth4dgPIflvvWbGo8yd",
            "x": Array [
              Object {
                "bytes": 911720,
                "context": Object {
                  "custom": Object {
                    "alt": "lady gazing to the right",
                  },
                },
                "created_at": "2022-08-16T06:59:26Z",
                "duration": null,
                "format": "png",
                "height": 826,
                "metadata": Array [],
                "original_secure_url": "https://res.cloudinary.com/kraft-heinz-whats-cooking-ca/image/upload/v1660633166/brand-experience/test-images/creative-content-panel/creative-content-panel-hero-banner-desktop_wvdfoa.png",
                "original_url": "http://res.cloudinary.com/kraft-heinz-whats-cooking-ca/image/upload/v1660633166/brand-experience/test-images/creative-content-panel/creative-content-panel-hero-banner-desktop_wvdfoa.png",
                "public_id": "brand-experience/test-images/creative-content-panel/creative-content-panel-hero-banner-desktop_wvdfoa",
                "raw_transformation": "f_auto/q_auto",
                "resource_type": "image",
                "secure_url": "https://res.cloudinary.com/kraft-heinz-whats-cooking-ca/image/upload/f_auto/q_auto/v1660633166/brand-experience/test-images/creative-content-panel/creative-content-panel-hero-banner-desktop_wvdfoa.png",
                "tags": Array [],
                "type": "upload",
                "url": "http://res.cloudinary.com/kraft-heinz-whats-cooking-ca/image/upload/f_auto/q_auto/v1660633166/brand-experience/test-images/creative-content-panel/creative-content-panel-hero-banner-desktop_wvdfoa.png",
                "version": 1660633166,
                "width": 1920,
              },
            ],
          },
        ],
      },
    ],
    "description": "Do you have isCool=true in your query params?",
    "distribution": Array [
      Object {
        "end": 0,
        "index": 0,
        "start": 0,
      },
      Object {
        "end": 1,
        "index": 1,
        "start": 0,
      },
    ],
    "id": "1zPgxFzQNPjhvjRKBPr203",
    "name": "are-you-cool",
    "sticky": false,
    "trafficAllocation": 1,
    "type": "nt_personalization",
  },
]
`;
